externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala

```diff
@@ -41,21 +41,30 @@ private[spark] final class SparkSQLEngine(name: String, spark: SparkSession)
 
   override private[kyuubi] val backendService = new SparkSQLBackendService(spark)
 
-  override protected def stopServer(): Unit = spark.stop()
+  override protected def stopServer(): Unit = {
+    spark.stop()
+    timeoutChecker.shutdown()
+    timeoutChecker.awaitTermination(10, TimeUnit.SECONDS)
+  }
 
   override def start(): Unit = {
     val interval = conf.get(KyuubiConf.ENGINE_CHECK_INTERVAL)
     val idleTimeout = conf.get(KyuubiConf.ENGINE_IDLE_TIMEOUT)
     val checkTask = new Runnable {
       override def run(): Unit = {
-        val current = System.currentTimeMillis
         val sessionManager = backendService.sessionManager.asInstanceOf[SparkSQLSessionManager]
-        if (sessionManager.getOpenSessionCount <= 0 &&
-          current - sessionManager.getLogoutTime > idleTimeout) {
+        val openSessionCount = sessionManager.getOpenSessionCount
+        val logoutTime = sessionManager.getLogoutTime
+        val current = System.currentTimeMillis
+        debug(s"timeoutChecker, openSessionCount:$openSessionCount," +
+          s" current time:$current, logoutTime:$logoutTime, idleTimeout:$idleTimeout")
+        if (openSessionCount <= 0 && (current - logoutTime) > idleTimeout) {
```

Comment:
Not only related to this PR but also the original one, there is a concurrent issue here. When we are closing a session, the `openSessionCount` could be set to 0, but the `logoutTime` here hasn't been updated, then the whole engine will self terminated unexpectedly.

Reply:
If we update the `logoutTime` first, then remove the `Session`?

Reply:
make sense